viperx/vars.go

```diff
@@ -20,8 +20,24 @@ func UnmarshalKey(key string, destination interface{}) error {
 	}
 
 	var b bytes.Buffer
-	if err := json.NewEncoder(&b).Encode(mapx.ToJSONMap(viper.Get(key))); err != nil {
+
+	// This may be a string in the case where a value was provided via an env var.
+	// If it's a string, try to decode it as json.
+	if v, ok := value.(string); ok {
+		if _, err := b.WriteString(v); err != nil {
+			return errors.WithStack(err)
+		}
+
+		// Try decoding the json directly
+		if err := json.NewDecoder(&b).Decode(destination); err == nil {
+			return nil
+		}
+	}
+
+	// If it's not a string or not valid json, use the value as it was originally provided
+	if err := json.NewEncoder(&b).Encode(mapx.ToJSONMap(value)); err != nil {
```

Comment:
So this looks good but is not that easy to read. Maybe you can change it to be more like:
```go
	switch v := value.(type) {
	case string:
		if _, err := b.WriteString(v); err != nil {
			return errors.WithStack(err)
		}
	default:
		if err := json.NewEncoder(&b).Encode(mapx.ToJSONMap(v)); err != nil {
			return errors.WithStack(err)
		}
	}
	// at this point the buffer holds the encoded value
	return errors.WithStack(json.NewDecoder(&b).Decode(destination))
```

Reply:
Sure thing. I had it as a `switch` before because I was trying a slightly different version but changed it because a switch with only one case and a default felt weird, but I don't mind switching it back if you think it's more readable.

Reply:
Actually this won't quite work as written because it doesn't account for the fallback. If it's a string but it can't be parsed as JSON we still need to run the `json.NewEncoder(&b).Encode(mapx.ToJSONMap(v))` piece.

Reply:
OK, my idea was just whether we could move the JSON decode to the end of the function and only have it once for every branch of the program.

Reply:
I don't think so, because in the case of a string that is not JSON,  `Decode` actually does need to be called twice. Once to realize it's not JSON, and then again after the `Encode` happens.
There are really 3 cases that are being handled here: 
1. The value is of whatever non-string type came in from viper (probably a map) - encode then decode
2. The value is of type `string` and contains JSON - just decode the JSON
3. The value is of type `string` and does not contain JSON - encode then decode (this case seems a bit wasteful as we're just encoding and decoding a string, but it's also necessary because the `Encode` lets it fail properly if it's trying to unmarshal into the wrong type.)